internal/core/community.go

```diff
@@ -80,55 +88,88 @@ func New(ctx context.Context, conf *Config) (ret *Community, err error) {
 func (p *Community) Article(ctx context.Context, id string) (article *Article, err error) {
-	if id == "123" {
-		article = &Article{
-			ArticleEntry{
-				ID:    id,
-				Title: "Title",
-			},
-			contentSummary,
-		}
-		return
+	article = &Article{}
+	sqlStr := "select id,title,user_id,cover,tags,content,ctime,mtime from article where id=?"
+	err = p.db.QueryRow(sqlStr, id).Scan(&article.ID, &article.Title, &article.UId, &article.Cover, &article.Tags, &article.Content, &article.Ctime, &article.Mtime)
+	if err == sql.ErrNoRows {
+		log.Println("not found the article")
+		return article, ErrNotExist
+	} else if err != nil {
+		return article, err
 	}
-	return nil, ErrNotExist
+	// TODO add author info
+	return
 }
 
-// CanEditable
+// CanEditable determine whether the user has the permission to operate.
 func (p *Community) CanEditable(ctx context.Context, uid, id string) (editable bool, err error) {
+	sqlStr := "select id from article where id=? and user_id = ?"
+	err = p.db.QueryRow(sqlStr, id, uid).Scan(&id)
+	if err != nil {
+		return false, err
+	}
 	return true, nil
 }
 
 // PutArticle adds new article (ID == "") or edits an existing article (ID != "").
 func (p *Community) PutArticle(ctx context.Context, uid string, article *Article) (id string, err error) {
-	return
+	// new article
+	if article.ID == "" {
+		sqlStr := "insert into article (title, ctime, mtime, user_id, tags, cover, content) values (?, ?, ?, ?, ?, ?, ?)"
+		res, err := p.db.Exec(sqlStr, &article.Title, time.Now(), time.Now(), &article.UId, &article.Tags, &article.Cover, &article.Content)
+		if err != nil {
+			return "", err
+		}
+		idInt, err := res.LastInsertId()
+		return strconv.FormatInt(idInt, 10), nil
+	}
+	// edit article
+	// can uid edit the article
+	canEdit, err := p.CanEditable(ctx, uid, article.ID)
+	if !canEdit {
+		log.Println("no permissions")
+		return "", ErrPermission
+	}
+	sqlStr := "update article set title=?, mtime=?, tags=?, cover=?, content=? where id=?"
+	_, err = p.db.Exec(sqlStr, &article.Title, time.Now(), &article.Tags, &article.Cover, &article.Content, &article.ID)
+	return article.ID, err
 }
 
+// DeleteArticle delete the article.
 func (p *Community) DeleteArticle(ctx context.Context, uid, id string) (err error) {
+	// can uid delete the article
+	canEdit, err := p.CanEditable(ctx, uid, id)
+	if !canEdit {
+		log.Println("no permissions")
+		return ErrPermission
+	}
+	sqlStr := "delete from article where id=?"
+	_, err = p.db.Exec(sqlStr, id)
+	// TODO delete the media
 	return
 }
 
-const (
-	MarkBegin = ""
-	MarkEnd   = "eof"
-)
-
 // ListArticle lists articles from an position.
-func (p *Community) ListArticle(ctx context.Context, from string, limit int) (items []*ArticleEntry, next string, err error) {
-	if from == MarkBegin {
-		item := &ArticleEntry{
-			ID:    "123",
-			Title: "Title",
+func (p *Community) ListArticle(ctx context.Context, from int, limit int) (items []*ArticleEntry, next int, err error) {
```

Comment:
don't change func prototype. `from`, `next` should be `string` (not `int`) for implementation-independent.

Reply:
And all this api should provide unit tests.